neqo-crypto/src/agent.rs

```diff
@@ -51,6 +51,146 @@ use crate::{
+    /// Decoding was unsuccessful, for example, not enough memory
+    fn decode(input: &[u8], output: &mut [u8]) -> Res<usize>;
+}
+
+/// The trait is responsible for calling `CertificateCompression` encoding and decoding
+/// functions using the NSS types
+unsafe impl<T: CertificateCompressor> UnsafeCertCompression for T {
+    unsafe extern "C" fn decode_callback(
+        input: *const ssl::SECItem,
+        output: *mut ::std::os::raw::c_uchar,
+        output_len: usize,
+        used_len: *mut usize,
+    ) -> ssl::SECStatus {
+        unsafe {
+            match NonNull::new(input.cast_mut()) {
+                None => ssl::SECFailure,
+                Some(input) => {
+                    if input.as_ref().data.is_null() || input.as_ref().len == 0 {
+                        return ssl::SECFailure;
+                    }
+
+                    let input_slice = null_safe_slice(input.as_ref().data, input.as_ref().len);
+                    let output_slice: &mut [u8] =
+                        core::slice::from_raw_parts_mut(output, output_len);
+                    let decode_result = T::decode(input_slice, output_slice);
+                    match decode_result {
+                        Err(_) => ssl::SECFailure,
+                        Ok(decoded_len) => {
+                            if decoded_len != output_len {
+                                return ssl::SECFailure;
+                            }
+
+                            *used_len = decoded_len;
+                            ssl::SECSuccess
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    unsafe extern "C" fn encode_callback(
+        input: *const ssl::SECItem,
+        output: *mut ssl::SECItem,
+    ) -> ssl::SECStatus {
+        unsafe {
+            match NonNull::new(input.cast_mut()) {
+                None => ssl::SECFailure,
+                Some(input) => {
+                    if input.as_ref().data.is_null() || input.as_ref().len == 0 {
+                        return ssl::SECFailure;
+                    }
+
+                    let input_len = input.as_ref().len;
+
+                    p11::SECITEM_AllocItem(
+                        null_mut(),
+                        // p11::SECItem is the same as ssl::SECItem
+                        output.cast::<p11::SECItemStr>(),
+                        input_len + 1,
+                    );
+
+                    let input_slice = null_safe_slice(input.as_ref().data, input_len);
+                    let output_slice: &mut [u8] = core::slice::from_raw_parts_mut(
+                        (*output).data,
+                        (*output).len.try_into().unwrap(),
+                    );
+
+                    let encode_result = T::encode(input_slice, output_slice);
+                    match encode_result {
+                        Err(_) => ssl::SECFailure,
+                        Ok(encoded_len) => {
+                            if encoded_len == 0 {
```

Comment:
IMPORTANT: you need to check that `encoded_len <= output.len()`.  I would add a `debug_assert!()` for that in addition to a runtime check that returns SECFailure.